examples/rate_providers/rate_providers.py

```diff
@@ -0,0 +1,134 @@
+from golem_core.low.activity import Activity
+from golem_core.low.market import Proposal
+from golem_core.mid import (
+    Buffer,
+    Chain,
+    Limit,
+    Map,
+    default_create_activity,
+    default_create_agreement,
+    default_negotiate,
+)
+
+FRAME_CONFIG_TEMPLATE = json.loads(Path("frame_params.json").read_text())
+PAYLOAD = Payload.from_image_hash(
+    "9a3b5d67b0b27746283cb5f287c13eab1beaa12d92a9f536b747c7ae"
+)
+SUBNET = "public"
+
+# scores will contain time of profiling execution
+scores: Dict[str, Optional[timedelta]] = {}
+
+
+async def delay(delay: timedelta, func: Callable) -> Any:
+    await asyncio.sleep(delay.total_seconds())
+    return func()
+
+
+async def filter_rated_providers(offer_stream: AsyncIterator[Proposal]) -> AsyncIterator[Proposal]:
+    async for offer in offer_stream:
+        provider_id = offer.data.issuer_id
+        if provider_id is None:
+            continue
+        if provider_id not in scores:
+            scores[provider_id] = None
+            yield offer
+
+
+async def prepare_activity(activity: Activity) -> Activity:
+    print(f"Prepare {activity}")
+    batch = await activity.execute_commands(
+        commands.Deploy(),
+        commands.Start(),
+        commands.SendFile("cubes.blend", "/golem/resource/scene.blend"),
+    )
+    await batch.wait(timeout=120)
+    return activity
+
+
+async def rate_provider(activity: Activity) -> str:
+    provider_id = activity.parent.parent.data.issuer_id
+    assert provider_id is not None
+    print(f"Rating {provider_id}")
+    frame_config = FRAME_CONFIG_TEMPLATE.copy()
+    fname = f"out{0:04d}.png"  # format used by `run-blender.sh` included in `PAYLOAD`
+    output_file = f"{provider_id}-{fname}"
+
+    start_time = datetime.utcnow()
+    batch = await activity.execute_commands(
+        commands.Run(f"echo '{json.dumps(frame_config)}' > /golem/work/params.json"),
+        commands.Run(["/golem/entrypoints/run-blender.sh"]),
+        commands.DownloadFile(f"/golem/output/{fname}", output_file),
+    )
+    await batch.wait(timeout=20)
+
+    end_time = datetime.utcnow()
+    scores[provider_id] = end_time - start_time
+
+    return output_file
+
+
+async def on_exception(func: Callable, args: Tuple, e: Exception) -> None:
+    activity = args[0]
+    print(f"Activity {activity} failed because of:\n{e}")
+    await activity.parent.close_all()
+
+
+async def main() -> None:
+    golem = GolemNode()
+    golem.event_bus.listen(DefaultLogger().on_event)
+
+    async with golem:
+        allocation = await golem.create_allocation(1.0)
+        payment_manager = DefaultPaymentManager(golem, allocation)
+        demand = await golem.create_demand(
+            PAYLOAD, allocations=[allocation], subnet=SUBNET
+        )
+
+        # `set_no_more_children` has to be called so `initial_proposals` will eventually stop yielding
+        asyncio.create_task(delay(timedelta(seconds=5), demand.set_no_more_children))
```

Comment:
There are different solutions available to achieve something similar:
1. Just `break` out of the `async for result in chain` loop
2. Add a `Limit` somewhere
3. Wrap `demand.initial_proposals()` in something like
```
async def proposals(demand, end time):
    async for proposal in demand.initial_proposals():
        if datetime.now() >= end_time:
            return
        yield proposal
```

I'm curious why did you choose this particular solution?
I'm not saying it is wrong :) But I'm also not super-sure if `set_no_more_children` should be a default solution for such problems.

Reply:
And regarding the comment: `demand.initial_proposals()` yields only when a value is requested.
Once you have processed 10 providers nothing more will be requested, so nothing more will be yielded.